ModelUploader/Program.cs

```diff
@@ -168,70 +170,99 @@ private static bool UploadModel(string file)
+                        foreach (JsonElement item in extendsSection.EnumerateArray())
+                        {
+                            if (item.ValueKind == JsonValueKind.String)
+                            {
+                                exitProcess = exitProcess || ProcessExtendsSectionItem(file, item.GetString());
+                            }
+                            else
+                            {
+                                exitProcess = true;
+                                Log.Error(String.Format("Unexpected extends value type {0}", item.ValueKind));
+                            }
+                        }
+                        break;
+
+                    case JsonValueKind.String:
+                        exitProcess = ProcessExtendsSectionItem(file, extendsSection.GetString());
+                        break;
+
+                    default:
+                        exitProcess = true;
+                        Log.Error(String.Format("Unexpected extends value type {0}", extendsSection.ValueKind));
+                        break;
+                }
+            }
+           
+            return exitProcess;
+        }
+
+
+        private static bool ProcessExtendsSectionItem(string file, string extendsSectionItem)
+        {
+            bool exitProcess = false;
+            MatchCollection dtmiExtends = Regex.Matches(extendsSectionItem, "dtmi:[:a-zA-Z0-9;_]*", RegexOptions.Singleline);
 
-                foreach (Match dtmiExtendsMatch in dtmiExtends)
+            foreach (Match dtmiExtendsMatch in dtmiExtends)
+            {
+                // find this model
+                try
                 {
-                    // find this model
-                    try
-                    {
-                        Response<ModelData> res = client.GetModel(dtmiExtendsMatch.Value);
-                        // model found! keep going
-                    }
-                    catch (RequestFailedException e)
+                    Response<ModelData> res = client.GetModel(dtmiExtendsMatch.Value);
+                    // model found! keep going
+                }
+                catch (RequestFailedException e)
+                {
+                    // Model Not Found - find it in the directory and call Upload Model
+                    if (e.Status == 404)
                     {
-                        // Model Not Found - find it in the directory and call Upload Model
-                        if (e.Status == 404)
-                        {
-                            string missingInterface = dtmiExtendsMatch.Value;
-                            int missingPartIndex = missingInterface.LastIndexOf(":");
-                            missingInterface = missingInterface.Substring(missingPartIndex + 1).Replace(";1", "");
-                            missingInterface = missingInterface.Replace(";", "");
+                        string missingInterface = dtmiExtendsMatch.Value;
```

Comment:
Here it seems we assume that files will be named per the local name fragment of the DTMI. Is this a reasonable assumption? Of course, searching through the contents of the set of model files, as opposed to just their names, may be prohibitively expensive..